server/app/assets/javascripts/session.ts

```diff
@@ -0,0 +1,395 @@
+      action: () => this.handleTimeout(),
+    })
+
+    timeouts.push({
+      time: data.totalTimeout,
+      action: () => this.handleTimeout(),
+    })
+
+    // Sort by earliest time
+    timeouts.sort((a, b) => a.time - b.time)
+
+    // We will always have at least one timeout in the future
+    const nextTimeout = timeouts[0]
+    const delay = (nextTimeout.time - now) * 1000
+
+    this.timer = window.setTimeout(() => {
+      nextTimeout.action()
+
+      // Check for next timeout after handling this one
+      this.checkAndSetTimer()
+    }, delay)
+  }
+
+  /**
+   * Set up event handlers for the server-rendered modals and process the extend session form submission.
+   */
+  private static setupModalEventHandlers() {
+    // HTMX after-request handler for session extension
+    document.addEventListener('htmx:afterRequest', (event: Event) => {
+      const customEvent = event as CustomEvent
+      const detail = customEvent.detail as {
+        xhr: XMLHttpRequest
+        elt: HTMLElement
+      }
+      if (detail.elt.id !== 'extend-session-form') return
+
+      this.hideModal('session-inactivity-warning')
+      this.inactivityWarningShown = false
+
+      // Processes /extend-session form submissions
+      if (detail.xhr.status === 200) {
+        const successText =
+          document.getElementById('session-extended-success-text')
+            ?.textContent || 'Session successfully extended'
+        ToastController.showToastMessage({
+          id: 'session-extended-toast',
+          content: successText,
+          type: 'success',
+          duration: 3000,
+          canDismiss: true,
+          canIgnore: false,
+          condOnStorageKey: null,
+        })
+        this.checkAndSetTimer()
+      } else {
+        const errorText =
+          document.getElementById('session-extended-error-text')?.textContent ||
+          'Failed to extend session'
+        ToastController.showToastMessage({
+          id: 'session-extend-error-toast',
+          content: errorText,
+          type: 'error',
+          duration: 3000,
+          canDismiss: true,
+          canIgnore: false,
+          condOnStorageKey: null,
+        })
+      }
+    })
+
+    /**
+     * Handle modal button actions using event delegation:
+     * - Primary actions:
+     *   - For inactivity warning: Extends session by submitting form to /extend-session
+     *   - For total length warning: Initiates logout
+     * - Secondary/close actions:
+     *   - Hides the modal
+     *   - Resets warning visibility flags
+     *   - Rechecks timeout conditions
+     *
+     * Uses data attributes to identify buttons and their actions:
+     * - data-modal-type: Identifies which modal the button belongs to
+     * - data-modal-primary: Primary action button (extend/logout)
+     * - data-modal-secondary: Secondary action button (dismiss)
+     * - data-close-modal: Close button (X)
+     */
+    document.addEventListener('click', (event) => {
```

Comment:
question (maybe blocking): Is there a way to apply this click handler just to the modals instead of the entire document?

Reply:
Applying them separately to the different modals based on their ids could help simplify some of the if/else logic.

Reply:
Adding separate click handler for dialogs. Event delegation help simplify code and improve perf in more complex SPAs but in our case, maybe it was an overkill.

Reply:
Ah that's good to know!